[WebProfiler] Fixed AJAX requests not being displayed in toolbar if started before toolbar was initialized

Somrlik opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no 
| Tickets       | Fix #38331 
| License       | MIT

A pretty lengthy fix, but I don't see a better solution.

The global variables work by initializing to an empty array -> the proxies fill the array -> the array is processed by the toolbar -> the `push()` method of the arrays is overridden -> the toolbar continues to process calls to `push()` using a custom handler.

Since proxying of calls to `window.fetch` and `XMLHttpRequest.prototype.open` has to be done before any scripts are executed, I added them just before `</head>`. This complicates the `ToolbarListener` a bit and I don't really know how to properly test it.

Since `fetch` and `XMLHttpRequest` are wildly different, I chose to create two global stacks and process them separately. They could be unified and the code could be minified a little but I chose not to for the sake of clarity.

Finally, since the ticket talks about Axios, you can try that 

```html
<!DOCTYPE html>
<html lang="">
<head>
    <meta charset="UTF-8">
    <title>Axios demo</title>
</head>
<body>
    <script src="https://unpkg.com/axios/dist/axios.min.js"></script>

    <pre id="axios">
        Pending...
    </pre>

    <script type="application/javascript">
    axios.get('https://api.github.com/zen').then((r) => {
        document.getElementById('axios').innerText = r.data;
    }).catch((err) => {
        document.getElementById('axios').innerText = err;
    });
    </script>
</body>
</html>
```

doesn't display the request unless this fix is applied.

Possible TODOs
 - [ ] Better test for `WebDebugToolbarListener`
 - [ ] Some sort of a Selenium test...?

Somrlik committed:
[WebProfiler] Fixed AJAX requests not being displayed in toolbar if started before toolbar was initialized

nicolas-grekas commented on src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php:
```diff
@@ -117,6 +117,19 @@ public function onKernelResponse(ResponseEvent $event)
+            $requestStack = $this->twig->render(
+                '@WebProfiler/Profiler/request_stack_js.html.twig',
+                [
+                    'csp_script_nonce' => isset($nonces['csp_script_nonce']) ? $nonces['csp_script_nonce'] : null,
```
`$nonces['csp_script_nonce'] ?? null,`

btw, how is `csp_script_nonce` populated? Is it automatic? (I don't rememeber and I didn't check :grimacing: )

Somrlik replied:
[Using the `cspHandler`](https://github.com/symfony/symfony/blob/392f2e0ba2821288d305463033f835905c8f8bd3/src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php#L83)

`$nonces = $this->cspHandler ? $this->cspHandler->updateResponseHeaders($request, $response) : [];`

nicolas-grekas commented on src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php:
```diff
@@ -117,6 +117,19 @@ public function onKernelResponse(ResponseEvent $event)
+                    'csp_script_nonce' => isset($nonces['csp_script_nonce']) ? $nonces['csp_script_nonce'] : null,
+                ]
+            );
+            $content = substr($content, 0, $pos).$requestStack.substr($content, $pos);
```
we should use `substr_replace()` here

Somrlik replied:
@nicolas-grekas Why? We're not replacing anything, only adding

nicolas-grekas replied:
because that's more efficient, and totally fine to add something in the middle of another string

Somrlik replied:
@nicolas-grekas Ah, I see now, had to read the docs twice :smiley:

nicolas-grekas commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -269,21 +269,19 @@
-            if (window.fetch && window.fetch.polyfill === undefined) {
-                var oldFetch = window.fetch;
-                window.fetch = function () {
-                    var promise = oldFetch.apply(this, arguments);
```
we might want to still redefine `window.fetch()` when `</head>` is not found?

Somrlik replied:
@nicolas-grekas So check if the `window.Sfjs_globalFetchStack` is set and if not, redefine it?

nicolas-grekas replied:
I think so yes

Somrlik replied:
@nicolas-grekas OK, I added the proxying if the global stacks are not found

Somrlik committed:
[WebProfiler] CR Fixes

 - Replace two calls to substr with substr_replace
 - When the global request stacks are not found, fall back to processing the requests using a proxy registered on toolbar load

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/request_stack_js.html.twig:
```diff
@@ -0,0 +1,24 @@
+            proxied.apply(this, Array.prototype.slice.call(arguments));
+        };
+    }
+})();
```
what is the purpose of deferring the execution? Couldn't it be called directly?

Somrlik replied:
@jderusse I felt that it would be more readable when debugging a request call in console in chrome, but I haven't tested it in other browsers. A direct call without deferring might be a bit confusing when viewing the call stack, as when deffered like this it doesn't show up in the main stack.

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php:
```diff
@@ -117,6 +117,19 @@ public function onKernelResponse(ResponseEvent $event)
         $content = $response->getContent();
+        $pos = strripos($content, '</head>');
+
+        if (false !== $pos) {
```
I'd also inject the request_stak when the `</head>` is not present to reduce duplicate code (see my comments below)

Somrlik replied:
@jderusse Thats reasonable, but where else to inject the code? The issue is about the requests not being processed before the `WebProfiler` registers its handlers and the only way to register before other scripts is to register in `<head>`.

Sure, an HTML document is valid witout a `head` tag; this was already discussed [here](https://github.com/symfony/symfony/pull/38451#discussion_r500889267) and forced some code to be duplicated.

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -269,101 +269,132 @@
+                window.Sfjs_globalFetchStack.push = function (args) {
+                    processSingleFetch(args[0], args[1]);
                 };
+            } else if (window.fetch && window.fetch.polyfill === undefined) {
```
This code would not be needed if `request_stack_js.html.twig` is always injeted and executed before this

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -269,101 +269,132 @@
+                window.Sfjs_globalXHRStack.push = function (args) {
+                    processSingleXHR(args[0], args[1], args[2]);
+                };
+            } else if (window.XMLHttpRequest && XMLHttpRequest.prototype.addEventListener) {
```
same here, not needed with `request_stack_js.html.twig`

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -269,101 +269,132 @@
+                });
+
+                /* Rewrite the global fetch stack push method */
+                window.Sfjs_globalFetchStack.push = function (args) {
```
`window.Sfjs_globalFetchStack` could be empty to free memory. before overriding the method `push`

Somrlik replied:
@jderusse Oh, you mean emptying the stack to preserve memory. IMHO the few bytes will never get collected by the JS engine anyways and manually deleting the entries would be more cycles wasted. The current code doesn't delete the stack anyways.

But I'm open to improvements, should I delete them?

jderusse replied:
It's never emptied. That's why you could do it.

just before this line, from line 322 to 324, you iterate over the array to process messages stacked.
Then, this line is about overriding the "push" method to process messages in "real time".
The issue is: All messages stack before reaching line 322 will stay. 

what you can do:
```js
window.Sfjs_globalFetchStack.forEach(function (req) {
  processSingleFetch(req[0], req[1]);
});
window.Sfjs_globalFetchStack.length = 0; // free memory

window.Sfjs_globalFetchStack.push = function (args) {
  ...
})
```

Somrlik replied:
@jderusse OK

[docs link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/length)

> You can set the length property to truncate an array at any time.

jderusse commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -269,101 +269,132 @@
+                });
+
+                /* Rewrite the global xhr stack push method */
+                window.Sfjs_globalXHRStack.push = function (args) {
```
same here `window.Sfjs_globalXHRStack.length = 0` to free memory

Somrlik committed:
[WebProfiler] [CR] Truncate the global arrays to length=0

nicolas-grekas commented:
Sorry to ask that late @Somrlik but if you're still up to finish this PR, could you please have a look at @jderusse comments above and also rebase on 5.4?

nicolas-grekas commented:
I'm sorry but I'm going to close as staled here. Please anyone, resubmit when you can!
